garden_ai/utils/_meta.py

```diff
@@ -0,0 +1,144 @@
+import inspect
+import linecache
+import textwrap
+
+
+def redef_in_main(cls):
+    """Helper: redefine a class in __main__, e.g. garden_ai._Model -> __main__._Model.
+
+    This has the effect of coaxing dill into serializing both the definition and
+    the instance of this object together, so it can be deserialized without
+    needing the definition to be available for import on the other side.
+
+    This works because dill is smart enough to know that if you defined a class
+    interactively (like in a repl) then it can't expect that definition to be
+    available after the session exits, so has to include it.
+
+    The upshot is that we can embed stuff like the _Model class as an implicit part of
+    the "function" that we register with globus compute, so there won't be any
+    garden-specific dependencies users need to worry about in the container.
+    """
+
+    # make sure it's not already in main
+    if cls.__module__ != "__main__":
+        import __main__
+
+        s = inspect.getsource(cls)
+        exec(s, __main__.__dict__)
+
+
+def make_func_to_serialize(pipeline):
+    """
+    Ensure that the composed function we register with globus compute is
+    "mainified", so that it has no references to the `garden_ai` namespace.
+    """
+    import __main__
+
+    redef_in_main(garden_ai_compose_functions)
+
+    functions = [step.func for step in pipeline.steps]
+    return __main__.garden_ai_compose_functions(*functions)
+
+
+def garden_ai_compose_functions(*functions):
+    """helper: compose functions together and inject special `_env_vars` kwarg.
+
+    Defines other helper functions locally to prevent polluting the __main__
+    namespace (see: `make_func_to_serialize`).
+
+    Any future "pipeline middleware" (like `inject_env_kwarg`) should also be
+    defined locally inside this function.
+    """
+
+    def compose(*functions):
+        func, *funcs = functions
+
+        def inner(*args, **kwargs):
+            result = func(*args, **kwargs)
+            for f in funcs:
+                result = f(result)
+            return result
```

Comment:
I'm pretty into this. You can have as many args in the "entrypoint" to your pipeline, but the rest of the steps can take and return one thing. That seems easy to explain.